contracts/utils/structs/Vector.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+library Vector {
+    error Empty();
+    error OutOfBound();
+
+    struct Bytes32Vector {
+        int128 begin;
+        int128 end;
+        mapping(int128 => bytes32) data;
+    }
+
+    function pushBack(Bytes32Vector storage vector, bytes32 value) internal {
+        unchecked {
+            vector.data[vector.end++] = value;
+        }
+    }
+
+    function pushFront(Bytes32Vector storage vector, bytes32 value) internal {
+        unchecked {
+            vector.data[--vector.begin] = value;
+        }
+    }
+
+    function popFront(Bytes32Vector storage vector) internal returns (bytes32 value) {
+        if (empty(vector)) revert Empty();
+        unchecked {
+            int128 idx = vector.begin++;
+            value = vector.data[idx];
+            delete vector.data[idx];
+        }
+    }
+
+    function popBack(Bytes32Vector storage vector) internal returns (bytes32 value) {
+        if (empty(vector)) revert Empty();
+        unchecked {
+            int128 idx = --vector.end;
+            value = vector.data[idx];
+            delete vector.data[idx];
+        }
+    }
+
+    function front(Bytes32Vector storage vector) internal view returns (bytes32 value) {
+        if (empty(vector)) revert Empty();
+        unchecked {
+            return vector.data[vector.begin];
+        }
+    }
+
+    function back(Bytes32Vector storage vector) internal view returns (bytes32 value) {
+        if (empty(vector)) revert Empty();
+        unchecked {
+            return vector.data[vector.end - 1];
+        }
+    }
+
+    function at(Bytes32Vector storage vector, uint256 i) internal view returns (bytes32 value) {
+        // leave check here: overflow could happen
+        int128 idx = vector.begin + int128(int256(i));
+        if (idx >= vector.end) revert OutOfBound();
+        return vector.data[idx];
+    }
+
+    function clear(Bytes32Vector storage vector) internal {
+        vector.begin = 0;
+        vector.end = 0;
+    }
+
+    function length(Bytes32Vector storage vector) internal view returns (uint256) {
+        unchecked {
```

Comment:
`begin` and `end` are only updated by ++ or --. They are not expected to overflow (it would take 2**127 calls for it to overflow).

a consequence is that any checks (like the safemath that solidity includes by default and safecast) would just be wasted gas.

For example, on a `pushX`/`popX` operation, unchecked saves ~100 gas per operation.